programs/server/Server.cpp

```diff
@@ -1118,6 +1118,18 @@ int Server::main(const std::vector<std::string> & /*args*/)
             if (config->has("max_partition_size_to_drop"))
                 global_context->setMaxPartitionSizeToDrop(config->getUInt64("max_partition_size_to_drop"));
 
+            if (config->has("total_max_threads"))
```

Comment:
Maybe it is better to name it `max_threads_for_all_queries` (by analogy with `max_memory_usage_for_all_queries`, now deprecated) ?
Also worth adding an example into `config.xml` (or update documentation), otherwise how to know about this new setting?
Plus in your initial article there is an explanation for this config directive, worth put it in the comment/doc.
And definitely put a word about "at least one thread will be allowed anyway"

Reply:
@azat , we've failed to attract attention to the parameter name https://github.com/ClickHouse/ClickHouse/issues/36551#issuecomment-1126112304 and decided to go with _total_max_threads_
If you believe that max_threads_for_all_queries is better, we'll willingly switch to this name.

Reply:
@ilejn thanks for the note. I don't have strong feeling about this, but `max_threads_for_all_queries` looks more "in ClickHouse style" I guess.

Reply:
max_threads_for_all_queries -  it a good name. I also thought that it's more in CH-style. So I will rename it. 
The minor question is about naming the supporting functions:
```
    /// Get total number of threads for all queries in process list.
    size_t getTotalNumThreads() const;

    size_t getTotalMaxThreads() const { return total_max_threads; }

    void setTotalMaxThreads(size_t total_max_threads_)
```

What do you think should be the correct names? 

So you these these are good variants? 
```

    size_t getNumThreadsForAllQueries() const;

    size_t getMaxThreadsForAllQueries() const { return max_threads_for_all_queries; }

    void setMaxThreadsForAllQueries(size_t max_threads_for_all_queries_)
```

Reply:
To me, functions can be leased as now, since it is already attached to processes, so simply total_max_threads is fine.

Reply:
I did not understand - should I rename parameter  total_max_threads into max_threads_for_all_queries or not? (and in which places?). If we decide to rename then I would rename it everywhere to keep the same naming logic everywhere.

Reply:
>should I rename parameter total_max_threads into max_threads_for_all_queries or not?

I think it is more in ClickHouse style, maybe you should wait until someone else will comment on this.

Reply:
> Also worth adding an example into `config.xml` (or update documentation), otherwise how to know about this new setting?

@azat in the latest commit I added a documentation text into settings.md and also into config.xml please take a look. Also I added -1 as default value into config.xml to see how will CI tests will run. I will probably remove -1 and set 0 in future config.xml and remove all logging code after the pull request will be in finished form.